lib/features/build_server/build_provider.dart

```diff
@@ -0,0 +1,265 @@
+import 'package:flyde/core/networking/protocol/project_init.dart';
+import 'package:flyde/core/networking/server.dart';
+import 'package:flyde/core/networking/websockets/session.dart';
+import 'package:flyde/features/build_server/cache/cache.dart';
+import 'package:flyde/features/build_server/cache/project_cache.dart';
+import 'package:flyde/features/build_server/compiler_interface.dart';
+
+/// Provides the build service on the local network.
+///
+/// A call to `setup` is mandatory to ensure that connections can be accepted.
+class BuildProvider {
+  /// The server that is used to handle connections.
+  final WebServer _server;
+
+  /// The list of interfaces to isolates which handle compilation.
+  final Map<String, MainInterface> _interfaces = {};
+
+  final Map<String, String> _availableProjects = {};
+
+  /// The cache that is used to store the compiled code.
+  late final Cache _cache;
+
+  /// A queue for each project that manages which client can
+  /// access a project and which have to wait.
+  final Map<String, Queue<ServerSession>> _projectCapacityQueues = {};
+
+  BuildProvider(this._server);
+
+  /// Sets up the server.
+  Future<void> setup({Directory? cacheDirectory}) async {
+    await _server.ready;
+
+    _cache = await Cache.load(from: cacheDirectory);
+    _server.wsOnMessage = _handleWebSocketMessage;
+    _server.wsOnDone = (ServerSession session) {
+      final String? id = session.storage['id'];
+
+      if (id != null) {
+        _activateNextSession(id, removing: session);
+      }
+    };
+  }
+
+  /// Kills all project isolates and shuts down the server.
+  Future<void> terminate() async {
+    _availableProjects.keys.toList().forEach(kill);
+    _availableProjects.clear();
+    _interfaces.clear();
+    await _server.close();
+  }
+
+  /// Terminates the isolate which hosts the project with the id [projectId].
+  void kill(String projectId) {
+    final MainInterface? interface = _interfaces[projectId];
+
+    if (interface != null) {
+      interface.isolate.isolate.kill(priority: Isolate.immediate);
+    }
+
+    if (_availableProjects.containsKey(projectId)) {
+      _availableProjects.remove(projectId);
+    }
+  }
+
+  /// Returns a list of all projects which have an active connection.
+  List<String> get projectIds {
+    return _availableProjects.keys.toList();
+  }
+
+  /// Returns the name of the project with the [projectId].
+  ///
+  /// If the project is not available, an argument error is thrown.
+  String projectName(String projectId) {
+    final String? name = _availableProjects[projectId];
+
+    if (name != null) {
+      return name;
+    }
+
+    throw ArgumentError('Project with id $projectId does not exists');
+  }
+
+  /// Handles a new websocket message.
+  Future<void> _handleWebSocketMessage(ServerSession session, dynamic message) async {
+    if (message is ProjectInitRequest) {
+      await _handleProjectInit(session, message);
```

Comment:
Why are you awaiting all of them if they are 'fire and forget'? You could convert a lot of `Future<void>` to simple `void` return types. Only use await when timing matters.

Reply:
I'm awaiting them to potentially do some work afterwards. I intend to log the requests so I need to do work after the handlers have been called.

Reply:
-> Planning for the future, ok